File: 102-binary-tree-level-order-traversal/binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        Queue<Pair<TreeNode,Integer>> qt = new LinkedList<>();
        List<List<Integer>> ans = new ArrayList<>();
        if(root == null){
            return ans;
        }
        qt.offer(new Pair<>(root,0));
        while(!qt.isEmpty()){
            Pair<TreeNode, Integer> pair= qt.poll();
            TreeNode newnode = pair.getKey();
            int level = pair.getValue();

            if(ans.size() == level){
                ans.add(new ArrayList<>());
            }
            ans.get(level).add(newnode.val);
            if(newnode.left != null){
                qt.offer(new Pair<>(newnode.left,level+1));
            }
            if(newnode.right != null){
                qt.offer(new Pair<>(newnode.right,level+1));
            }
        }
        return ans;
    }
}

File: 102-binary-tree-level-order-traversal/binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if(root == NULL){
            return ans;
        }
        queue<pair<TreeNode*,int>> qt;
        qt.push({root,0});
        while(!qt.empty()){
            auto [node,level] = qt.front();
            qt.pop();
            if(ans.size() == level){
                ans.push_back({});
            }
            ans[level].push_back(node->val);
            if(node->left != NULL){
                qt.push({node->left,level+1});
            }
            if(node->right != NULL){
                qt.push({node->right,level+1});
            }
        }
        return ans;
    }
};